src/smt/theory_str_noodler/counter_automaton.h

```diff
@@ -0,0 +1,178 @@
+#ifndef Z3_STR_COUNTER_AUT_H_
+#define Z3_STR_COUNTER_AUT_H_
+
+#include <iostream>
+#include <map>
+#include <set>
+#include <queue>
+#include <string>
+#include <memory>
+#include <concepts>
+#include <compare>
+
+#include <mata/nfa/nfa.hh>
+#include <mata/nfa/strings.hh>
+#include <mata/nfa/builder.hh>
+
+#include "formula.h"
+#include "util.h"
+
+namespace smt::noodler::ca {
+
+    /**
+     * @brief Structured alphabet. Performs mapping between mata symbols 
+     * and particular symbols T.
+     * 
+     * @tparam T Symbol type
+     */
+    template <typename T>
+    requires std::strict_weak_order<std::less<T>, T const&, T const&>
+    class StructAlphabet {
+
+    private:
+        std::map<T, mata::Symbol> alph_symb_mata {};
+        std::map<mata::Symbol, T> alph_mata_symb {};
+
+    public:
+
+        StructAlphabet() : alph_symb_mata(), alph_mata_symb() { }
+
+        mata::Symbol add_symbol(const T& symb) {
+            mata::Symbol new_symbol = this->alph_symb_mata.size();
+            auto iter = this->alph_symb_mata.find(symb);
+            if (iter != this->alph_symb_mata.end()) {
+                return iter->second;
+            } else {
+                this->alph_symb_mata[symb] = new_symbol;
+                this->alph_mata_symb[new_symbol] = symb;
+                return new_symbol;
+            }
+        };
+
+        void insert(const mata::Symbol mata_symb, const T& symb) {
+            this->alph_symb_mata[symb] = mata_symb;
+            this->alph_mata_symb[mata_symb] = symb;
+        }
+
+        mata::Symbol get_mata_symbol(const T& symb) const {
+            return this->alph_symb_mata.at(symb);
+        }
+
+        const T& get_symbol(const mata::Symbol symb) const {
+            return this->alph_mata_symb.at(symb);
+        }
+
+        std::set<T> get_all_symbols() const {
+            std::set<T> ret;
+            for (const auto& [key,val] : this->alph_symb_mata) {
+                ret.insert(key);
+            }
+            return ret;
+        }
+
+    };
+
+    /**
+     * @brief Symbols of the form <mark, var, label, symbol>
+     */
+    struct AtomicSymbol {
+        char mark; // 0 = L; 1 = P; 2 = R
+        BasicTerm var; // variable from string constraint
+        char label; // 0 = missing value
+        mata::Symbol symbol; // symbol (used for the R-case)
```

Comment:
What is the value if mark != R?

Reply:
It is not defined as such a symbol should not occur.

Reply:
I mean, mark can be L, so there will be some symbol, is it going to be undefined? Or does R-case means something else, not the value of the mark?

Reply:
I made it clear through the constructor functions setting default values to unused fields.

Reply:
If you want to make it even safer, you can make the default constructor private, so you can only create AtomicSymbol using one of those three functions.

Reply:
Good idea, I made the constructor private.